sw/device/lib/dif/dif_flash_ctrl.c

```diff
@@ -0,0 +1,1046 @@
+
+#include "sw/device/lib/dif/dif_flash_ctrl.h"
+
+#include "dif_warn_unused_result.h"
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/mmio.h"
+
+#include "flash_ctrl_regs.h"
+
+// Helper macros for statically checking register layout assumptions.
+#define FLASH_CTRL_EXPAND_REGION(id, name, suffix) \
+  FLASH_CTRL_MP_REGION_CFG_##id##_##name##_##id##suffix
+
+#define FLASH_CTRL_ASSERT_REGION(name, suffix, id1, id2)          \
+  _Static_assert(FLASH_CTRL_EXPAND_REGION(id1, name, suffix) ==   \
+                     FLASH_CTRL_EXPAND_REGION(id2, name, suffix), \
+                 "Region layout changed.")
+
+#define FLASH_CTRL_ASSERT_REGION_UNIFORMITY(name, suffix) \
+  FLASH_CTRL_ASSERT_REGION(name, suffix, 0, 1);           \
+  FLASH_CTRL_ASSERT_REGION(name, suffix, 1, 2);           \
+  FLASH_CTRL_ASSERT_REGION(name, suffix, 2, 3);           \
+  FLASH_CTRL_ASSERT_REGION(name, suffix, 3, 4);           \
+  FLASH_CTRL_ASSERT_REGION(name, suffix, 4, 5);           \
+  FLASH_CTRL_ASSERT_REGION(name, suffix, 5, 6);           \
+  FLASH_CTRL_ASSERT_REGION(name, suffix, 6, 7)
+
+// Helper macro to check that region registers are contiguous.
+#define FLASH_CTRL_ASSERT_REGION_CONTINUITY(id1, id2)             \
+  _Static_assert(FLASH_CTRL_MP_REGION_CFG_##id1##_REG_OFFSET +    \
+                         FLASH_CTRL_PARAM_REG_WIDTH / 8 ==        \
+                     FLASH_CTRL_MP_REGION_CFG_##id2##_REG_OFFSET, \
+                 "Region register layout is not contiguous.")
+
+// Helper macros only check the first 8 regions and must be updated if regions
+// are added.
+_Static_assert(FLASH_CTRL_PARAM_NUMREGIONS == 8,
+               "Total number of regions changed.");
+
+// Static assert all region config registers are laid out the same. If any of
+// these checks fail it could mean using `FLASH_CTRL_MP_REGION_CFG_0_*` bit
+// offsets as common offsets across all MP regions is invalid.
+// TODO Remove this when MP region common parameters are added.
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(EN, );
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(RD_EN, );
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(PROG_EN, );
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(ERASE_EN, );
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(BASE, _MASK);
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(BASE, _OFFSET);
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(SIZE, _MASK);
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(SIZE, _OFFSET);
+FLASH_CTRL_ASSERT_REGION_UNIFORMITY(PARTITION, );
+
+// Static assert register layout is contiguous. This is necessary to ensure
+// `region_config_offset_get()` provides valid register offsets.
+FLASH_CTRL_ASSERT_REGION_CONTINUITY(0, 1);
+FLASH_CTRL_ASSERT_REGION_CONTINUITY(1, 2);
+FLASH_CTRL_ASSERT_REGION_CONTINUITY(2, 3);
+FLASH_CTRL_ASSERT_REGION_CONTINUITY(3, 4);
+FLASH_CTRL_ASSERT_REGION_CONTINUITY(4, 5);
+FLASH_CTRL_ASSERT_REGION_CONTINUITY(5, 6);
+FLASH_CTRL_ASSERT_REGION_CONTINUITY(6, 7);
+
+const static bitfield_field32_t kControlOpBitfield = {
```

Comment:
just a note here that the latest flash change adds scramble / ecc to the equation as well.
i'm still working on updating the doc.